### PR TITLE
[MXNET-33] Enhance mkldnn pooling to support full convention

### PR DESCRIPTION
**Pull request by TaoLv**
## Description ##
Follow-up for https://github.com/apache/incubator-mxnet/pull/10021

Then gluon/squeezenet can benefit from this PR with below performance improvement.

Squeezenet1.1 inference on SKX8180:

BS | Before | After | Speed-up
-- | -- | -- | --|
1 | 114.891003 | 265.844219 | 2.313881958
2 | 155.3841 | 429.361312 | 2.76322553
4 | 182.232015 | 687.565177 | 3.773020767
8 | 203.664481 | 845.273893 | 4.15032552
16 | 221.299678 | 1110.591241 | 5.018494609
32 | 232.902484 | 1497.973739 | 6.431763686
64 | 232.521016 | 1464.001852 | 6.296213036
128 | 234.921866 | 1285.150863 | 5.470545952
256 | 243.123314 | 1152.503337 | 4.740406496

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by pengzhao-intel** [comment]
Mark to get the update email

File: src/operator/nn/mkldnn/mkldnn_pooling-inl.h [hunk 1]
```diff
@@ -87,23 +87,8 @@ inline bool SupportMKLDNNPooling(const PoolingParam &param) {
...
-
-// need to support pooling convention full
-// https://issues.apache.org/jira/browse/MXNET-33
-#if 0
-  if (((dshape[2] + 2 * param.pad[0] - param.kernel[0]) % param.stride[0] == 0) &&
-      ((dshape[3] + 2 * param.pad[1] - param.kernel[1]) % param.stride[1] == 0))
-    return true;
-  else
-    return false;
-#endif
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_pooling-inl.h]: can we ignore the shape completely? even for the case of kValid?

**Comment by zheng-da** [comment]
can you write unit tests for this?

Could you also explain what is the difference between full and valid? I know it has something to do with padding, but I don't remember the details.

**Comment by TaoLv** [comment]
@zheng-da I notice there already has pooling test case for both kValid and kFull: https://github.com/apache/incubator-mxnet/blob/master/tests/python/gpu/test_operator_gpu.py#L558
Do you think we still another test cast for this PR?

Pooling output shape is calculated with: 
```
o_h = (i_h +pad_top + pad_bottom - k_h) / stride_h ) + 1
o_w = (i_w +pad_left + pad_right - k_w) / stride_w ) + 1
```
So if `i_h +pad_top + pad_bottom - k_h` is not divisible by `stride_h`, there will be two possible `o_h`. One is by rounding up (ceil, kFull) and the other is by rounding down (floor, kValid).
In mkldnn, output shape is provided by user/developer during creating primitive/descriptor. mkldnn will check the provided output shape according to the given input shape, padding size, stride size, kernel size.

In mkldnn pooling operator, we need adjust (increase) padding size (especially for pad_right and pad_bottom) to make sure the output shape (calculated by infershape) matches those input shape, padding size, stride size and kernel size.

**Comment by zheng-da** [comment]
the test can't really cover the cases we want. it didn't detect the bug before @ashokei disabled full convention.
you can change its stride to (3, 3) or something to make sure `i_h +pad_top + pad_bottom - k_h` can't be divided by `stride_h `.

**Comment by TaoLv** [comment]
@zheng-da I see. Will add test case later.

File: src/operator/nn/mkldnn/mkldnn_pooling.cc [hunk 2]
```diff
@@ -150,6 +150,16 @@ mkldnn::pooling_forward::primitive_desc GetPoolingFwd(const PoolingParam &param,
   int pad_l_ = param.pad[1], pad_r_ = param.pad[1];
   int stride_h_ = param.stride[0], stride_w_ = param.stride[1];
 
+  if (param.pooling_convention == pool_enum::kFull) {
```
Comment (pengzhao-intel) [inline src/operator/nn/mkldnn/mkldnn_pooling.cc]: Is it possible to write up a macro/function for the same check of `kFull` in the different place in this file?

Comment (TaoLv) [reply src/operator/nn/mkldnn/mkldnn_pooling.cc]: Done.

**Comment by TaoLv** [comment]
@marcoabreu please help to review the new test case.

**Comment by pengzhao-intel** [comment]
@piiswrong Could help merge this PR? thanks

File: tests/python/gpu/test_operator_gpu.py [hunk 3]
```diff
@@ -920,6 +920,35 @@ def test_3d_pooling(pool_type, p_value=2):
...
+
+        # o_h = ceil((10 + 1 + 1 - 4) / 3) + 1 = 4
+        # o_w = ceil((10 + 2 + 2 - 5) / 4) + 1 = 4
+        ctx_list.append({'ctx': mx.cpu(0), 'pool_data': data, 'type_dict': {'pool_data': np.float32}})
+        sym_list.append(mx.sym.Pooling(kernel=kernel, pad=pad, stride=stride, pool_type=pool_type,
+                                       pooling_convention=convention, global_pool=True, name='pool'))
+
+        ctx_list.append({'ctx': mx.gpu(0), 'pool_data': data, 'type_dict': {'pool_data': np.float32}})
+        sym_list.append(mx.sym.Pooling(kernel=kernel, pad=pad, stride=stride, pool_type=pool_type,
+                                       pooling_convention=convention, global_pool=True, name='pool'))
```
Comment (zheng-da) [inline tests/python/gpu/test_operator_gpu.py]: it doesn't seem you test your code. Once global_pool is true, all paddings are set to 0.

Comment (zheng-da) [reply tests/python/gpu/test_operator_gpu.py]: Please check your test code with and without your fix to make sure that your test can trigger the bug.

**Comment by TaoLv** [comment]
I change the status of this PR to WIP since I'm confirming some designs of mkldnn pooling primitive. Will update this PR later.

**Comment by TaoLv** [comment]
The unit test is fixed. Also, MKL-DNN fixed an issue for full convention max-pooling and the fix is released in v0.15. This PR's merge should pend on the integration of MKL-DNN v0.15 release into MXNet. Meanwhile, I fallback the full convention avg-pooling to mxnet original implementation since there still has some known issues for avg-pooling in MKL-DNN. 
@zheng-da Are you OK for this? Or do you know any real models are using full convention avg-pooling? We need that information to decide/prioritize how to fix the issue in MKL-DNN.

**Comment by lupesko** [comment]
@zheng-da bouncing.
Adding @azai91 and @mseth10 for review.

**Comment by stu1130** [comment]
@zheng-da bouncing again.
@mxnet-label-bot [pr-awaiting-review]

**Comment by TaoLv** [comment]
@stu1130 Thank you for looking into this. We are still working on integrating higher versions of MKL-DNN into MXNet. There are some issues need be fixed for the integration. This PR is pending on that.

**Comment by vandanavk** [comment]
@sandeep-krishnamurthy Please change status to pr-work-in-progress

**Comment by Roshrini** [comment]
@TaoLv Any update on this PR? Can you resolve conflicts?

**Comment by TaoLv** [comment]
@Roshrini Thanks for asking. Once #12953 is merged, I can continue to work on this.

**Comment by ankkhedia** [comment]
@TaoLv Could you please look into CI failures for #12953 so that it moves fast and this PR can be taken forward?

**Comment by TaoLv** [comment]
@ankkhedia Code is rebased and ready for review.

**Comment by ankkhedia** [comment]
@azai91 @mseth10 ping for review!

**Comment by TaoLv** [comment]
@ankkhedia Is this good to go?

**Comment by ankkhedia** [comment]
@azai91 @mseth10 ping for review!

**Comment by kalyc** [comment]
@mxnet-label-bot update [pr-awaiting-review]

File: src/operator/nn/mkldnn/mkldnn_pooling-inl.h [hunk 4]
```diff
@@ -113,20 +113,15 @@ inline bool SupportMKLDNNPooling(const PoolingParam &param,
...
+  } else {
+    // currently, only max-pooling is supported for full convention
+    if (param.pool_type == pool_enum::kMaxPooling)
+      return true;
+    else
+      return false;
```
Comment (szha) [inline src/operator/nn/mkldnn/mkldnn_pooling-inl.h]: `return param.pool_type == pool_enum::kMaxPooling`

Comment (TaoLv) [reply src/operator/nn/mkldnn/mkldnn_pooling-inl.h]: Fixed. Thanks for review.  :)

**Pull request merged**